src/core/model.jl

```diff
@@ -190,13 +194,73 @@ Return an iterator over all agents of the model.
 """
 allagents(model) = values(model.agents)
 
-
 """
     allids(model)
 Return an iterator over all agent IDs of the model.
 """
 allids(model) = keys(model.agents)
 
+#######################################################################################
+# %% Higher order collections
+#######################################################################################
+export group_agent_iter, group_id_iter, group_agent_map, group_id_map
+
+"""
+    group_id_iter(order::Int, model::ABM)
+
+Return an iterator over all agent IDs of the model, group by order. When `order = 2`, the
+iterator returns agent pairs, and when `order = 3`: agent triples.
+
+See also [`group_agent_iter`](@ref).
+"""
+group_id_iter(order::Int, model::ABM) =
+    Iterators.product((by_id(model) for _ in 1:order)...)
+
+"""
+    group_agent_iter(order::Int, model::ABM)
+
+Return an iterator over all agents of the model, group by order. When `order = 2`, the
+iterator returns ID pairs, e.g `(1, 2)` and when `order = 3`: ID triples, e.g.
+`(1, 7, 8)`.
+
+See also [`group_id_iter`](@ref).
+"""
+group_agent_iter(order::Int, model::ABM) =
+    Iterators.product((map(i -> model[i], by_id(model)) for _ in 1:order)...)
+
+"""
+    group_id_map(order::Int, f::Function, model::ABM)
+    group_id_map(order::Int, f::Function, model::ABM, filter::Function)
+
+Applies function `f` to all grouped agent IDs of a [`group_id_iter`](@ref). `f` must take
+the form `f(model, a, b)`, where the number of arguments (after `model`) is equal to
+`order`.
+Optionally, a `filter` function that accepts an iterable and returns a `Bool` can be
+applied to remove unwanted matches from the results.
+
+See also [`group_agent_map`](@ref).
+"""
+group_id_map(order::Int, f::Function, model::ABM) =
+    (f(model, idx...) for idx in group_id_iter(order, model))
+group_id_map(order::Int, f::Function, model::ABM, filter::Function) =
+    (f(model, idx...) for idx in group_id_iter(order, model) if filter(idx))
+
+"""
+    group_agent_map(order::Int, f::Function, model::ABM)
+    group_agent_map(order::Int, f::Function, model::ABM, filter::Function)
+
+Applies function `f` to all grouped agents of a [`group_agent_iter`](@ref). `f` must take
+the form `f(a, b)`, where the number of arguments is equal to `order`.
+Optionally, a `filter` function that accepts an iterable and returns a `Bool` can be
+applied to remove unwanted matches from the results.
+
+See also [`group_agent_map`](@ref).
+"""
+group_agent_map(order::Int, f::Function, model::ABM) =
+    (f(idx...) for idx in group_agent_iter(order, model))
+group_agent_map(order::Int, f::Function, model::ABM, filter::Function) =
+    (f(idx...) for idx in group_agent_iter(order, model) if filter(idx))
```

Comment:
the mapping function `f` and the filtering function `filter` accept the arguments in a different way, which I find unintuitive. One requires `f(a, b, c)` while the other `f((a, b, c))`. They should have same input syntax.